applications/MedApplication/custom_utilities/med_testing_utilities.cpp

```diff
@@ -0,0 +1,197 @@
+}
+
+// check if the geometries have the same nodes
+// for this to be the case, two conditions need to be fulflled:
+// 1: geometries have same number of nodes
+// 2: nodes of both geometries are in same order and in same coords
+bool have_same_nodes(
+    const GeometryType& rGeom1,
+    const GeometryType& rGeom2)
+{
+    if (rGeom1.PointsNumber() != rGeom2.PointsNumber()) return false;
+
+    for (std::size_t i=0; i<rGeom1.PointsNumber(); ++i) {
+        if (rGeom1[i].Distance(rGeom2[i]) > 1e-12) return false;
+    }
+
+    return true;
+}
+
+void CheckEntitiesAreEqual(
+    const NodeType& rNode1,
+    const NodeType& rNode2)
+{
+    KRATOS_TRY
+
+    // KRATOS_CHECK_EQUAL(rNode1.Id(), rNode2.Id()); // the MEDApp deliberately does not care about IDs
+
+    KRATOS_CHECK_DOUBLE_EQUAL(rNode1.X(),  rNode2.X());
+    KRATOS_CHECK_DOUBLE_EQUAL(rNode1.X0(), rNode2.X0());
+
+    KRATOS_CHECK_DOUBLE_EQUAL(rNode1.Y(),  rNode2.Y());
+    KRATOS_CHECK_DOUBLE_EQUAL(rNode1.Y0(), rNode2.Y0());
+
+    KRATOS_CHECK_DOUBLE_EQUAL(rNode1.Z(),  rNode2.Z());
+    KRATOS_CHECK_DOUBLE_EQUAL(rNode1.Z0(), rNode2.Z0());
+
+    KRATOS_CATCH("")
+}
+
+void CheckGeometriesAreEqual(
+    const GeometryType& rGeom1,
+    const GeometryType& rGeom2)
+{
+    KRATOS_TRY
+
+    // KRATOS_CHECK_EQUAL(rGeom1.Id(), rGeom2.Id()); // the MEDApp deliberately does not care about IDs
+
+    KRATOS_CHECK_EQUAL(rGeom1.PointsNumber(), rGeom2.PointsNumber());
+
+    KRATOS_CHECK(GeometryType::IsSame(rGeom1, rGeom2));
+
+    for (std::size_t i=0; i<rGeom1.PointsNumber(); ++i) {
+        CheckEntitiesAreEqual(rGeom1[i], rGeom2[i]);
+    }
+
+    KRATOS_CATCH("")
+}
+
+template<class TContainerType>
+void CheckEntitiesAreEqual(
+    const TContainerType& rEntities1,
+    const TContainerType& rEntities2)
+{
+    KRATOS_TRY
+
+    // basic checks
+    KRATOS_CHECK_EQUAL(rEntities1.size(), rEntities2.size());
+
+    // check entities
+    for (std::size_t i=0; i<rEntities1.size(); ++i) {
+        CheckEntitiesAreEqual(*(rEntities1.begin()+i), *(rEntities2.begin()+i));
+    }
+
+    KRATOS_CATCH("")
+}
+
+void CheckGeometriesAreEqual(
+    const ModelPart& rModelPart1,
+    const ModelPart& rModelPart2)
+{
+    KRATOS_TRY
+
+    // basic checks
+    KRATOS_CHECK_EQUAL(rModelPart1.NumberOfGeometries(), rModelPart2.NumberOfGeometries());
+
+    auto check_geoms = [](const ModelPart& rModelPart1, const ModelPart& rModelPart2){
+        for (auto geom_1 : rModelPart1.Geometries()) {
+            for (auto geom_2 : rModelPart2.Geometries()) {
+                if (have_same_nodes(geom_1, geom_2)) {
+                    CheckGeometriesAreEqual(geom_1, geom_2);
+                    goto here;
+                }
+            }
+            KRATOS_ERROR << "no match found for geometry " << geom_1 << std::endl;
+            here:;
+        }
+    };
+
+    check_geoms(rModelPart1, rModelPart2);
+    check_geoms(rModelPart2, rModelPart1);
```

Comment:
it seems to me that `have_same_nodes` and `CheckGeometriesAreEqual` are commutative, so the second check (`check_geom`(rModelPart2, rModelPart`);`) is unnecessary.

Reply:
not sure

ModelPart1 could have less geometries than ModelPart2, which would only be detected in the second call

Checking if the geometries are the same in a modelpart is quite nasty :/

Reply:
true, but you're already checking whether the two `ModelPart`s have the same number of geometries in line 121. Anyway, performance in a test isn't that important so this is just a minor comment.

Reply:
hm good point

I will think about this again next time I develop, I remember that I thought long about this